src/util/thread_pool.h

```diff
@@ -0,0 +1,210 @@
+/*
+Minetest
+Copyright (C) 2010-2013 celeron55, Perttu Ahola <celeron55@gmail.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#ifndef UTIL_THREAD_POOL_HEADER
+#define UTIL_THREAD_POOL_HEADER
+
+#include "../threading/mutex.h"
+#include "../threading/mutex_auto_lock.h"
+#include "../threading/semaphore.h"
+#include "../threading/thread.h"
+
+class PoolableUpdateThread : public Thread
+{
+public:
+	PoolableUpdateThread(const std::string &name, Semaphore *sem,
+		Mutex *quitcount_mutex, u32 *quitcount) :
+		Thread(name + "Update"),
+		m_sem(sem),
+		m_quitcount_mutex(quitcount_mutex),
+		m_quitcount(quitcount) {}
+	~PoolableUpdateThread() {}
+
+	void *run()
+	{
+		DSTACK(FUNCTION_NAME);
+		BEGIN_DEBUG_EXCEPTION_HANDLER
+
+		while (!stopRequested()) {
+			m_sem->wait();
+
+			if (stopRequested())
+				break;
+
+			{
+				MutexAutoLock lock(*m_quitcount_mutex);
+				if (m_quitcount < 0) {
+					--m_quitcount;
+					break;
+				}
+			}
+
+			doUpdate();
+		}
+
+		END_DEBUG_EXCEPTION_HANDLER
+
+		return NULL;
+	}
+
+protected:
+	virtual void doUpdate() = 0;
+
+private:
+	Semaphore *m_sem;
+	Mutex *m_quitcount_mutex;
+	u32 *m_quitcount;
+};
+
+class ThreadPool
+{
+public:
+	ThreadPool(const std::string &name, size_t thread_count = 1) :
```

Comment:
`size_t` is a proper type. `u8` would go against scalability (even though I admit reaching 255+ threads is unlikely), and wouldn't strictly stick to the size type `std::vector` uses. `u8` as well would not improve performance (due to how integers of a bit size less than the native CPU mode word size are handled) nor memory usage (to to 1. non-conditional use by `std::vector` and 2. alignment rules).